File: mini-nn-java/src/main/java/com/mininn/Loss.java
package com.mininn;

/**
 * Loss functions for neural networks.
 */
public enum Loss {
    MSE,
    BINARY_CROSS_ENTROPY,
    CROSS_ENTROPY;

    private static final double EPSILON = 1e-7;

    public double compute(Tensor predictions, Tensor targets) {
        int n = predictions.getRows();
        double loss = 0;

        switch (this) {
            case MSE:
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < predictions.getCols(); j++) {
                        double diff = predictions.get(i, j) - targets.get(i, j);
                        loss += diff * diff;
                    }
                }
                return loss / (2.0 * n);

            case BINARY_CROSS_ENTROPY:
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < predictions.getCols(); j++) {
                        double p = clip(predictions.get(i, j));
                        double t = targets.get(i, j);
                        loss -= t * Math.log(p) + (1 - t) * Math.log(1 - p);
                    }
                }
                return loss / n;

            case CROSS_ENTROPY:
                for (int i = 0; i < n; i++) {
                    for (int j = 0; j < predictions.getCols(); j++) {
                        double p = clip(predictions.get(i, j));
                        double t = targets.get(i, j);
                        if (t > 0) {
                            loss -= t * Math.log(p);
                        }
                    }
                }
                return loss / n;

            default:
                return 0;
        }
    }

    public Tensor gradient(Tensor predictions, Tensor targets) {
        int rows = predictions.getRows();
        int cols = predictions.getCols();
        Tensor grad = new Tensor(rows, cols);

        switch (this) {
            case MSE:
                for (int i = 0; i < rows; i++) {
                    for (int j = 0; j < cols; j++) {
                        grad.set(i, j, (predictions.get(i, j) - targets.get(i, j)) / rows);
                    }
                }
                break;

            case BINARY_CROSS_ENTROPY:
                for (int i = 0; i < rows; i++) {
                    for (int j = 0; j < cols; j++) {
                        double p = clip(predictions.get(i, j));
                        double t = targets.get(i, j);
                        grad.set(i, j, (-t / p + (1 - t) / (1 - p)) / rows);
                    }
                }
                break;

            case CROSS_ENTROPY:
                // For softmax + cross-entropy, gradient simplifies to (pred - target)
                for (int i = 0; i < rows; i++) {
                    for (int j = 0; j < cols; j++) {
                        grad.set(i, j, (predictions.get(i, j) - targets.get(i, j)) / rows);
                    }
                }
                break;
        }

        return grad;
    }

    private static double clip(double v) {
        return Math.max(EPSILON, Math.min(1 - EPSILON, v));
    }
}

File: mini-nn-python/mini_nn/loss.py
"""Loss functions for neural networks."""

import numpy as np
from enum import Enum, auto
from .tensor import Tensor


class LossType(Enum):
    """Types of loss functions."""
    MSE = auto()
    BINARY_CROSS_ENTROPY = auto()
    CROSS_ENTROPY = auto()


class Loss:
    """Loss function implementations."""

    EPSILON = 1e-7

    @staticmethod
    def compute(predictions: Tensor, targets: Tensor, loss_type: LossType) -> float:
        """Compute loss value."""
        pred = predictions.numpy()
        targ = targets.numpy()
        n = pred.shape[0]

        if loss_type == LossType.MSE:
            return float(np.sum((pred - targ) ** 2) / (2 * n))

        elif loss_type == LossType.BINARY_CROSS_ENTROPY:
            p = np.clip(pred, Loss.EPSILON, 1 - Loss.EPSILON)
            return float(-np.sum(targ * np.log(p) + (1 - targ) * np.log(1 - p)) / n)

        elif loss_type == LossType.CROSS_ENTROPY:
            p = np.clip(pred, Loss.EPSILON, 1 - Loss.EPSILON)
            return float(-np.sum(targ * np.log(p)) / n)

        return 0.0

    @staticmethod
    def gradient(predictions: Tensor, targets: Tensor, loss_type: LossType) -> Tensor:
        """Compute gradient of loss with respect to predictions."""
        pred = predictions.numpy()
        targ = targets.numpy()
        n = pred.shape[0]

        if loss_type == LossType.MSE:
            return Tensor((pred - targ) / n)

        elif loss_type == LossType.BINARY_CROSS_ENTROPY:
            p = np.clip(pred, Loss.EPSILON, 1 - Loss.EPSILON)
            return Tensor((-targ / p + (1 - targ) / (1 - p)) / n)

        elif loss_type == LossType.CROSS_ENTROPY:
            # For softmax + cross-entropy, gradient simplifies to (pred - target)
            return Tensor((pred - targ) / n)

        return Tensor.zeros(pred.shape[0], pred.shape[1])
